### PR TITLE
Feature lua state per connection

### PR DESCRIPTION
**Pull request by Icety**

### CODE REVIEW

File: scripter/lua/lua.go [hunk 1]
```diff
@@ -18,75 +21,89 @@ var (
...
+	l.connections = map[string]scripterConn{}
 
-	return s, nil
+	return l, nil
 }
 
 // The scripter state to which scripter functions are attached
 type luaScripter struct {
 	name string
+	service string
```
Comment (martijnjanssen) [inline scripter/lua/lua.go]: There are multiple services per luaScripter, how does this work?

File: scripter/lua/lua.go [hunk 2]
```diff
@@ -18,75 +21,90 @@ var (
...
-// Get all scripts for a given service and pass the string to each script
-func (l *luaScripter) Handle(message string) (string, error) {
-	result := message
-	var err error
+//Return a connection for the given ip-address, if no connection exists yet, create it.
+func (l *luaScripter) GetConnection(service string, conn net.Conn) scripter.ConnectionWrapper {
+	s := strings.Split(conn.RemoteAddr().String(), ":")
+	s = s[:len(s)-1]
+	ip := strings.Join(s, ":")
+	log.Infof("Getting connection for: %s", ip)
```
Comment (martijnjanssen) [inline scripter/lua/lua.go]: Should be a debug message.

File: scripter/lua/lua.go [hunk 3]
```diff
@@ -18,75 +21,90 @@ var (
...
+	var sConn scripterConn
+	var ok bool
+
+	if sConn, ok = l.connections[ip]; !ok {
+		sConn = scripterConn{}
+		sConn.conn = conn
+		sConn.scripts = map[string]map[string]*lua.LState{}
+		sConn.cancelFuncs = map[string]map[string]context.CancelFunc{}
+		l.connections[ip] = sConn
+		log.Infof("New connection returned")
```
Comment (martijnjanssen) [inline scripter/lua/lua.go]: Debug log again.

File: scripter/lua/lua.go [hunk 4]
```diff
@@ -18,75 +21,90 @@ var (
...
+	}
 
-	// TODO: Figure out the correct way to call all handle methods
-	for _, ls := range l.scripts[l.name] {
-		result, err = handleScript(ls, result)
-		if err != nil {
-			return "", err
-		}
+	if !sConn.hasScripts(service) {
+		log.Infof("Adding scripts for service: %s", service)
```
Comment (martijnjanssen) [inline scripter/lua/lua.go]: Debug log.

File: scripter/lua/lua.go [hunk 5]
```diff
@@ -106,10 +121,49 @@ func handleScript(ls *lua.LState, message string) (string, error) {
 	return result, nil
 }
 
+//////////  Connection Wrapper struct \\\\\\\\\\\
```
Comment (nordinvannes) [inline scripter/lua/lua.go]: Thomas, what is this?

File: scripter/lua/lua.go [hunk 6]
```diff
@@ -106,10 +121,49 @@ func handleScript(ls *lua.LState, message string) (string, error) {
...
+}
+
+//Set a string function for a connection
+func (w *ConnectionStruct) SetStringFunction(name string, getString func() string) error {
+	return w.conn.SetStringFunction(name, getString, w.service)
+}
+
+
+
+//////////  Scripter Connection struct \\\\\\\\\\\
```
Comment (nordinvannes) [inline scripter/lua/lua.go]: What is this?

**Review by nordinvannes — APPROVED** [review]
Comment changes can be done. LGTM!

File: scripter/lua/lua.go [hunk 7]
```diff
@@ -118,7 +172,21 @@ func (l *luaScripter) SetStringFunction(name string, getString func() string) er
...
-func (l *luaScripter) Close() {
-	l.Close()
+//Returns if the scripts for a given service are loaded already
+func (c *scripterConn) hasScripts(service string) bool {
+	_, ok := c.scripts[service]
+	return ok
 }
+
+//Add scripts to a connection for a given service
+func (c *scripterConn) addScripts(service string, scripts map[string]string) {
```
Comment (martijnjanssen) [inline scripter/lua/lua.go:180]: It cannot be the case that none of these methods return an error.

**Pull request merged**